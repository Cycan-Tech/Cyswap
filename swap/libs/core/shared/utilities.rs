#![allow(non_snake_case)]
use primitives::U256;

pub fn encodePriceSqrt(reserve1: U256, reserve0: U256) -> U256 {
    // return BigNumber.from(
    //     new bn(reserve1.toString())
    //       .div(reserve0.toString())
    //       .sqrt()
    //       .multipliedBy(new bn(2).pow(96))
    //       .integerValue(3)
    //       .toString()
    //   )
    reserve1 / reserve0.integer_sqrt() * (U256::from(2).pow(U256::from(96)))
}

// export function expandTo18Decimals(n: number): BigNumber {
//     return BigNumber.from(n).mul(BigNumber.from(10).pow(18))
//   }

pub fn expandTo18Decimals(n:u128)->u128{
    n*(10u128.pow(18))
}

#[cfg(test)]
mod test{
    use super::expandTo18Decimals;

    #[test]
    fn tesTexpandTo18Decimals(){
        let result = expandTo18Decimals(2);
        println!("result is:{}",result);
        println!("u128 max is:{}",u128::MAX);
    }
}