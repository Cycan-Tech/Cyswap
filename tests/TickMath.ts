import BN from 'bn.js'
import { BigNumber } from 'ethers'
import { expect } from './shared/expect'
import { artifacts, network, patract } from 'redspot';
import Decimal from 'decimal.js';
import { encodePriceSqrt, MIN_SQRT_RATIO, MAX_SQRT_RATIO } from './shared/utilities';

const { getContractFactory, getRandomSigner } = patract;

const { api, getAddresses, getSigners } = network;

import { setupContract, fromSigner, setupProxy } from './helpers'

const MIN_TICK = -887272
const MAX_TICK = 887272

describe('TickMath', () => {
    after(() => {
        return api.disconnect();
    });

    // async function setup() {
    //     // await api.isReady;
    //     // const signerAddresses = await getAddresses();
    //     // const Alice = signerAddresses[0];
    //     // const sender = await getRandomSigner(Alice, '20000 UNIT');
    //     // const contractFactory = await getContractFactory('weth9_contract', sender.address);
    //     // const contract = await contractFactory.deploy('new', 'None','None');
    //     // // const abi = artifacts.readArtifact('metadata');
    //     // const receiver = await getRandomSigner();
    //     // return { sender, contractFactory, contract, receiver, Alice };
    // }

    async function setup() {
        const ian = await getRandomSigner()
        const kayne = await getRandomSigner()
        let contract = await setupContract('TickMathTest', 'new',);

        return { contract, kayne, ian }
    }

    describe('#getSqrtRatioAtTick', () => {
        it('throws for too low', async () => {
            await api.isReady;
            const { query: tickMathQuery, tx: tickMathTx, } = await setupContract("TickMathTest", "new");
            await expect(tickMathQuery.getSqrtRatioAtTick(MIN_TICK - 1)).to.eventually.be.rejectedWith('T');
        })

        it('throws for too low', async () => {
            const { query: tickMathQuery, tx: tickMathTx, } = await setupContract("TickMathTest", "new");
            await expect(tickMathQuery.getSqrtRatioAtTick(MAX_TICK + 1)).to.eventually.be.rejectedWith('T');
        })

        it('min tick', async () => {
            const { query: tickMathQuery, tx: tickMathTx, } = await setupContract("TickMathTest", "new");
            expect(await (await tickMathQuery.getSqrtRatioAtTick(MIN_TICK)).output).to.eq('4295128739')
        })

        it('min tick +1', async () => {
            const { query: tickMathQuery, tx: tickMathTx, } = await setupContract("TickMathTest", "new");
            expect(await (await tickMathQuery.getSqrtRatioAtTick(MIN_TICK + 1)).output).to.eq('4295343490')
        })

        it('max tick - 1', async () => {
            const { query: tickMathQuery, tx: tickMathTx, } = await setupContract("TickMathTest", "new");
            expect(await (await tickMathQuery.getSqrtRatioAtTick(MAX_TICK - 1)).output).to.eq('1461373636630004318706518188784493106690254656249')
        })

        it('max tick', async () => {
            const { query: tickMathQuery, tx: tickMathTx, } = await setupContract("TickMathTest", "new");
            expect(await (await tickMathQuery.getSqrtRatioAtTick(MAX_TICK)).output).to.eq('1461446703485210103287273052203988822378723970342')
        })

        it('min tick ratio is less than js implementation', async () => {
            const { query: tickMathQuery, tx: tickMathTx, } = await setupContract("TickMathTest", "new");
            let output = (await tickMathQuery.getSqrtRatioAtTick(MIN_TICK)).output;
            let ticker = Number.parseInt(output ? output.toString() : "0");
            expect(ticker < (encodePriceSqrt(1, BigNumber.from(2).pow(127))).toNumber());
        })

        it('max tick ratio is greater than js implementation', async () => {
            const { query: tickMathQuery, tx: tickMathTx, } = await setupContract("TickMathTest", "new");
            expect((await tickMathQuery.getSqrtRatioAtTick(MAX_TICK)).output).to.be.gt(encodePriceSqrt(BigNumber.from(2).pow(127), 1))
        })

        // for (const absTick of [
        //     50,
        //     100,
        //     250,
        //     500,
        //     1_000,
        //     2_500,
        //     3_000,
        //     4_000,
        //     5_000,
        //     50_000,
        //     150_000,
        //     250_000,
        //     500_000,
        //     738_203,
        // ]) {
        //     for (const tick of [-absTick, absTick]) {
        //         describe(`tick ${tick}`, () => {
        //             it('is at most off by 1/100th of a bips', async () => {
        //                 const { query: tickMathQuery, tx: tickMathTx, } = await setupContract("TickMathTest", "new");
        //                 const jsResult = new Decimal(1.0001).pow(tick).sqrt().mul(new Decimal(2).pow(96))
        //                 const result = await tickMathQuery.getSqrtRatioAtTick(tick)
        //                 const absDiff = new Decimal(result.output?result.output.toString():"0").sub(jsResult).abs()
        //                 expect(absDiff.div(jsResult).toNumber()).to.be.lt(0.000001)
        //             })
        //             // it('result', async () => {
        //             //   expect((await tickMathQuery.getSqrtRatioAtTick(tick)).toString()).to.matchSnapshot()
        //             // })
        //             // it('gas', async () => {
        //             //   await snapshotGasCost(tickMathQuery.getGasCostOfGetSqrtRatioAtTick(tick))
        //             // })
        //         })
        //     }
        // }


        describe('#MIN_SQRT_RATIO', async () => {
            it('equals #getSqrtRatioAtTick(MIN_TICK)', async () => {
                const { query: tickMathQuery, tx: tickMathTx, } = await setupContract("TickMathTest", "new");
                const min = await tickMathQuery.getSqrtRatioAtTick(MIN_TICK);
                expect(min.output).to.equal(await (await tickMathQuery.minSqrtRatio()).output)
                expect(min.output).to.equal(MIN_SQRT_RATIO)
            })
        })

        describe('#MAX_SQRT_RATIO', async () => {
            it('equals #getSqrtRatioAtTick(MAX_TICK)', async () => {
                const { query: tickMathQuery, tx: tickMathTx, } = await setupContract("TickMathTest", "new");
                const max = await tickMathQuery.getSqrtRatioAtTick(MAX_TICK)
                expect(max.output).to.equal(await (await tickMathQuery.maxSqrtRatio()).output)
                expect(max.output).to.equal(MAX_SQRT_RATIO)
            })
        })

    });

    describe('#getTickAtSqrtRatio', () => {
        it('throws for too low', async () => {
            const { query: tickMathQuery, tx: tickMathTx, } = await setupContract("TickMathTest", "new");
            await expect(tickMathQuery.getTickAtSqrtRatio(MIN_SQRT_RATIO.sub(1))).to.eventually.be.rejectedWith('R');
        })

        it('throws for too high', async () => {
            const { query: tickMathQuery, tx: tickMathTx, } = await setupContract("TickMathTest", "new");
            await expect(tickMathQuery.getTickAtSqrtRatio(BigNumber.from(MAX_SQRT_RATIO))).to.be.rejectedWith('R')
        })

        it('ratio of min tick', async () => {
            const { query: tickMathQuery, tx: tickMathTx, } = await setupContract("TickMathTest", "new");
            expect((await tickMathQuery.getTickAtSqrtRatio(MIN_SQRT_RATIO)).output).to.eq(MIN_TICK)
        })
        it('ratio of min tick + 1', async () => {
            const { query: tickMathQuery, tx: tickMathTx, } = await setupContract("TickMathTest", "new");
            expect((await tickMathQuery.getTickAtSqrtRatio('4295343490')).output).to.eq(MIN_TICK + 1)
        })
        it('ratio of max tick - 1', async () => {
            const { query: tickMathQuery, tx: tickMathTx, } = await setupContract("TickMathTest", "new");
            expect((await tickMathQuery.getTickAtSqrtRatio('1461373636630004318706518188784493106690254656249')).output).to.eq(MAX_TICK - 1)
        })
        it('ratio closest to max tick', async () => {
            const { query: tickMathQuery, tx: tickMathTx, } = await setupContract("TickMathTest", "new");
            expect((await tickMathQuery.getTickAtSqrtRatio(MAX_SQRT_RATIO.sub(1))).output).to.eq(MAX_TICK - 1)
        })

        for (const ratio of [
            MIN_SQRT_RATIO,
            encodePriceSqrt(BigNumber.from(10).pow(12), 1),
            encodePriceSqrt(BigNumber.from(10).pow(6), 1),
            encodePriceSqrt(1, 64),
            encodePriceSqrt(1, 8),
            encodePriceSqrt(1, 2),
            encodePriceSqrt(1, 1),
            encodePriceSqrt(2, 1),
            encodePriceSqrt(8, 1),
            encodePriceSqrt(64, 1),
            encodePriceSqrt(1, BigNumber.from(10).pow(6)),
            encodePriceSqrt(1, BigNumber.from(10).pow(12)),
            MAX_SQRT_RATIO.sub(1),
        ]) {
            describe(`ratio ${ratio}`, () => {
                it('is at most off by 1', async () => {
                    const { query: tickMathQuery, tx: tickMathTx, } = await setupContract("TickMathTest", "new");
                    const jsResult = new Decimal(ratio.toString()).div(new Decimal(2).pow(96)).pow(2).log(1.0001).floor()
                    const result = await (await tickMathQuery.getTickAtSqrtRatio(ratio)).output;
                    const absDiff = new Decimal(result ? result.toString() : "0").sub(jsResult).abs();
                    expect(absDiff.toNumber()).to.be.lte(1)
                })
                it('ratio is between the tick and tick+1', async () => {
                    const { query: tickMathQuery, tx: tickMathTx, } = await setupContract("TickMathTest", "new");
                    const tick = await (await tickMathQuery.getTickAtSqrtRatio(ratio)).output;
                    console.log("tick is----------:", new Decimal(tick ? tick.toString() : "0"));
                    const ratioOfTick = (await tickMathQuery.getSqrtRatioAtTick(tick ? tick.toString() : "0")).output
                    const ratioOfTickPlusOne = await (await tickMathQuery.getSqrtRatioAtTick(Number.parseInt(tick ? tick.toString() : "0") + 1)).output
                    console.log("ratio is-------", ratio.toString());
                    console.log("ratioOfTick is----------:", ratioOfTick?.toString());
                    console.log("ratioOfTickPlusOne is----------:", BigNumber.from(ratioOfTickPlusOne ? ratioOfTickPlusOne.toString() : "0"));
                    //TODO 79228162514264337593543 is error!
                    expect(new BN(ratio.toString())).to.be.gte(new BN(ratioOfTick ? ratioOfTick.toString() : "0"));

                    let ratioOf = new BN(ratioOfTickPlusOne ? ratioOfTickPlusOne.toString() : "0");
                    expect(new BN(ratio.toString())).to.be.lt(ratioOf);
                })
                it('result', async () => {
                    const { query: tickMathQuery, tx: tickMathTx, } = await setupContract("TickMathTest", "new");
                  expect(await tickMathQuery.getTickAtSqrtRatio(ratio)).to.matchSnapshot()
                })
                it('gas', async () => {
                    const { query: tickMathQuery, tx: tickMathTx, } = await setupContract("TickMathTest", "new");
                  expect(await tickMathQuery.getGasCostOfGetTickAtSqrtRatio(ratio)).to.matchSnapshot();
                })
            })
        }
    });
})