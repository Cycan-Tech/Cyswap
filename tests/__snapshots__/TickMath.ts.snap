// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TickMath #getTickAtSqrtRatio ratio 4295128739 gas 1`] = `
Object {
  "debugMessage": "",
  "gasConsumed": 19829139552,
  "gasRequired": 74999922688,
  "output": 4257995563,
  "result": Object {
    "ok": Object {
      "data": "0x2bdbcbfd00000000",
      "flags": Array [],
    },
  },
  "storageDeposit": Object {
    "charge": 0,
  },
}
`;

exports[`TickMath #getTickAtSqrtRatio ratio 4295128739 result 1`] = `
Object {
  "debugMessage": "",
  "gasConsumed": 19822795599,
  "gasRequired": 74999922688,
  "output": -887272,
  "result": Object {
    "ok": Object {
      "data": "0x1876f2ff",
      "flags": Array [],
    },
  },
  "storageDeposit": Object {
    "charge": 0,
  },
}
`;

exports[`TickMath #getTickAtSqrtRatio ratio 79228162514264337593543 gas 1`] = `
Object {
  "debugMessage": "",
  "gasConsumed": 19564494748,
  "gasRequired": 74999922688,
  "output": 3993350759,
  "result": Object {
    "ok": Object {
      "data": "0x67b205ee00000000",
      "flags": Array [],
    },
  },
  "storageDeposit": Object {
    "charge": 0,
  },
}
`;

exports[`TickMath #getTickAtSqrtRatio ratio 79228162514264337593543 result 1`] = `
Object {
  "debugMessage": "",
  "gasConsumed": 19558150795,
  "gasRequired": 74999922688,
  "output": -276324,
  "result": Object {
    "ok": Object {
      "data": "0x9cc8fbff",
      "flags": Array [],
    },
  },
  "storageDeposit": Object {
    "charge": 0,
  },
}
`;

exports[`TickMath #getTickAtSqrtRatio ratio 79228162514264337593543950 gas 1`] = `
Object {
  "debugMessage": "",
  "gasConsumed": 19464262538,
  "gasRequired": 74999922688,
  "output": 3893118549,
  "result": Object {
    "ok": Object {
      "data": "0x55460ce800000000",
      "flags": Array [],
    },
  },
  "storageDeposit": Object {
    "charge": 0,
  },
}
`;

exports[`TickMath #getTickAtSqrtRatio ratio 79228162514264337593543950 result 1`] = `
Object {
  "debugMessage": "",
  "gasConsumed": 19457918585,
  "gasRequired": 74999922688,
  "output": -138162,
  "result": Object {
    "ok": Object {
      "data": "0x4ee4fdff",
      "flags": Array [],
    },
  },
  "storageDeposit": Object {
    "charge": 0,
  },
}
`;

exports[`TickMath #getTickAtSqrtRatio ratio 9903520314283042199192993792 gas 1`] = `
Object {
  "debugMessage": "",
  "gasConsumed": 18395900832,
  "gasRequired": 74999922688,
  "output": 2824756843,
  "result": Object {
    "ok": Object {
      "data": "0x6b5e5ea800000000",
      "flags": Array [],
    },
  },
  "storageDeposit": Object {
    "charge": 0,
  },
}
`;

exports[`TickMath #getTickAtSqrtRatio ratio 9903520314283042199192993792 result 1`] = `
Object {
  "debugMessage": "",
  "gasConsumed": 18389556879,
  "gasRequired": 74999922688,
  "output": -41590,
  "result": Object {
    "ok": Object {
      "data": "0x8a5dffff",
      "flags": Array [],
    },
  },
  "storageDeposit": Object {
    "charge": 0,
  },
}
`;

exports[`TickMath #getTickAtSqrtRatio ratio 28011385487393069959365969113 gas 1`] = `
Object {
  "debugMessage": "",
  "gasConsumed": 19380621596,
  "gasRequired": 74999922688,
  "output": 3809477607,
  "result": Object {
    "ok": Object {
      "data": "0xe70310e300000000",
      "flags": Array [],
    },
  },
  "storageDeposit": Object {
    "charge": 0,
  },
}
`;

exports[`TickMath #getTickAtSqrtRatio ratio 28011385487393069959365969113 result 1`] = `
Object {
  "debugMessage": "",
  "gasConsumed": 19374277643,
  "gasRequired": 74999922688,
  "output": -20796,
  "result": Object {
    "ok": Object {
      "data": "0xc4aeffff",
      "flags": Array [],
    },
  },
  "storageDeposit": Object {
    "charge": 0,
  },
}
`;

exports[`TickMath #getTickAtSqrtRatio ratio 56022770974786139918731938227 gas 1`] = `
Object {
  "debugMessage": "",
  "gasConsumed": 19277707503,
  "gasRequired": 74999922688,
  "output": 3706563514,
  "result": Object {
    "ok": Object {
      "data": "0xbaabeddc00000000",
      "flags": Array [],
    },
  },
  "storageDeposit": Object {
    "charge": 0,
  },
}
`;

exports[`TickMath #getTickAtSqrtRatio ratio 56022770974786139918731938227 result 1`] = `
Object {
  "debugMessage": "",
  "gasConsumed": 19271363550,
  "gasRequired": 74999922688,
  "output": -6932,
  "result": Object {
    "ok": Object {
      "data": "0xece4ffff",
      "flags": Array [],
    },
  },
  "storageDeposit": Object {
    "charge": 0,
  },
}
`;

exports[`TickMath #getTickAtSqrtRatio ratio 79228162514264337593543950336 gas 1`] = `
Object {
  "debugMessage": "",
  "gasConsumed": 18140197472,
  "gasRequired": 74999922688,
  "output": 2569053483,
  "result": Object {
    "ok": Object {
      "data": "0x2ba5209900000000",
      "flags": Array [],
    },
  },
  "storageDeposit": Object {
    "charge": 0,
  },
}
`;

exports[`TickMath #getTickAtSqrtRatio ratio 79228162514264337593543950336 result 1`] = `
Object {
  "debugMessage": "",
  "gasConsumed": 18133853519,
  "gasRequired": 74999922688,
  "output": 0,
  "result": Object {
    "ok": Object {
      "data": "0x00000000",
      "flags": Array [],
    },
  },
  "storageDeposit": Object {
    "charge": 0,
  },
}
`;

exports[`TickMath #getTickAtSqrtRatio ratio 112045541949572279837463876454 gas 1`] = `
Object {
  "debugMessage": "",
  "gasConsumed": 19220035554,
  "gasRequired": 74999922688,
  "output": 3648891565,
  "result": Object {
    "ok": Object {
      "data": "0xadaa7dd900000000",
      "flags": Array [],
    },
  },
  "storageDeposit": Object {
    "charge": 0,
  },
}
`;

exports[`TickMath #getTickAtSqrtRatio ratio 112045541949572279837463876454 result 1`] = `
Object {
  "debugMessage": "",
  "gasConsumed": 19213691601,
  "gasRequired": 74999922688,
  "output": 6931,
  "result": Object {
    "ok": Object {
      "data": "0x131b0000",
      "flags": Array [],
    },
  },
  "storageDeposit": Object {
    "charge": 0,
  },
}
`;

exports[`TickMath #getTickAtSqrtRatio ratio 224091083899144559674927752909 gas 1`] = `
Object {
  "debugMessage": "",
  "gasConsumed": 19324408844,
  "gasRequired": 74999922688,
  "output": 3753264855,
  "result": Object {
    "ok": Object {
      "data": "0xd746b6df00000000",
      "flags": Array [],
    },
  },
  "storageDeposit": Object {
    "charge": 0,
  },
}
`;

exports[`TickMath #getTickAtSqrtRatio ratio 224091083899144559674927752909 result 1`] = `
Object {
  "debugMessage": "",
  "gasConsumed": 19318064891,
  "gasRequired": 74999922688,
  "output": 20795,
  "result": Object {
    "ok": Object {
      "data": "0x3b510000",
      "flags": Array [],
    },
  },
  "storageDeposit": Object {
    "charge": 0,
  },
}
`;

exports[`TickMath #getTickAtSqrtRatio ratio 633825300114114700748351602688 gas 1`] = `
Object {
  "debugMessage": "",
  "gasConsumed": 19397974198,
  "gasRequired": 74999922688,
  "output": 3826830209,
  "result": Object {
    "ok": Object {
      "data": "0x81cb18e400000000",
      "flags": Array [],
    },
  },
  "storageDeposit": Object {
    "charge": 0,
  },
}
`;

exports[`TickMath #getTickAtSqrtRatio ratio 633825300114114700748351602688 result 1`] = `
Object {
  "debugMessage": "",
  "gasConsumed": 19391630245,
  "gasRequired": 74999922688,
  "output": 41590,
  "result": Object {
    "ok": Object {
      "data": "0x76a20000",
      "flags": Array [],
    },
  },
  "storageDeposit": Object {
    "charge": 0,
  },
}
`;

exports[`TickMath #getTickAtSqrtRatio ratio 79228162514264337593543950336000 gas 1`] = `
Object {
  "debugMessage": "",
  "gasConsumed": 19620256604,
  "gasRequired": 74999922688,
  "output": 4049112615,
  "result": Object {
    "ok": Object {
      "data": "0x278e58f100000000",
      "flags": Array [],
    },
  },
  "storageDeposit": Object {
    "charge": 0,
  },
}
`;

exports[`TickMath #getTickAtSqrtRatio ratio 79228162514264337593543950336000 result 1`] = `
Object {
  "debugMessage": "",
  "gasConsumed": 19613912651,
  "gasRequired": 74999922688,
  "output": 138162,
  "result": Object {
    "ok": Object {
      "data": "0xb21b0200",
      "flags": Array [],
    },
  },
  "storageDeposit": Object {
    "charge": 0,
  },
}
`;

exports[`TickMath #getTickAtSqrtRatio ratio 79228162514264337593543950336000000 gas 1`] = `
Object {
  "debugMessage": "",
  "gasConsumed": 19610361194,
  "gasRequired": 74999922688,
  "output": 4039217205,
  "result": Object {
    "ok": Object {
      "data": "0x3590c1f000000000",
      "flags": Array [],
    },
  },
  "storageDeposit": Object {
    "charge": 0,
  },
}
`;

exports[`TickMath #getTickAtSqrtRatio ratio 79228162514264337593543950336000000 result 1`] = `
Object {
  "debugMessage": "",
  "gasConsumed": 19604017241,
  "gasRequired": 74999922688,
  "output": 276324,
  "result": Object {
    "ok": Object {
      "data": "0x64370400",
      "flags": Array [],
    },
  },
  "storageDeposit": Object {
    "charge": 0,
  },
}
`;

exports[`TickMath #getTickAtSqrtRatio ratio 1461446703485210103287273052203988822378723970341 gas 1`] = `
Object {
  "debugMessage": "",
  "gasConsumed": 19867827286,
  "gasRequired": 74999922688,
  "output": 4296683297,
  "result": Object {
    "ok": Object {
      "data": "0x212f1a0001000000",
      "flags": Array [],
    },
  },
  "storageDeposit": Object {
    "charge": 0,
  },
}
`;

exports[`TickMath #getTickAtSqrtRatio ratio 1461446703485210103287273052203988822378723970341 result 1`] = `
Object {
  "debugMessage": "",
  "gasConsumed": 19861483333,
  "gasRequired": 74999922688,
  "output": 887271,
  "result": Object {
    "ok": Object {
      "data": "0xe7890d00",
      "flags": Array [],
    },
  },
  "storageDeposit": Object {
    "charge": 0,
  },
}
`;
